File: Python3/1161_maximum_level_sum_of_a_binary_tree.py
# https://leetcode.cn/problems/maximum-level-sum-of-a-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        res, depth = 1, 1
        s_max = root.val
        q = deque()
        q.append(root)
        while q:
            n = len(q)
            s_curr = 0
            for _ in range(n):
                node = q.popleft()
                s_curr += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            if s_curr > s_max:
                res = depth
                s_max = s_curr
            depth += 1
        return res

File: Cpp/1161_maximum_level_sum_of_a_binary_tree.cpp
// https://leetcode.cn/problems/maximum-level-sum-of-a-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <deque>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxLevelSum(TreeNode* root) {
        int res = 1, depth = 1;
        long s_max = root->val;
        deque<TreeNode*> q;
        q.push_back(root);
        while (!q.empty()) {
            int n = q.size();
            long s_curr = 0;
            for (int i = 0; i < n; ++i) {
                auto node = q.front();
                q.pop_front();
                s_curr += node->val;
                if (node->left) q.push_back(node->left);
                if (node->right) q.push_back(node->right);
            }
            if (s_curr > s_max) {
                res = depth;
                s_max = s_curr;
            }
            depth += 1;
        }
        return res;
    }
};